File: android/src/services/CoachService.ts
import { getIntervalsIcuCredentials } from './ApiIntervalsIcu';
import systmSample from '../data/systm-sample.json';

// Coach readiness — the Android counterpart of desktop/src/services/coachservice.cpp
// (André, 2026-08-26: "port everything to android").
//
// One deliberate difference, and it is an IMPROVEMENT rather than a shortcut: the desktop
// computes CTL/ATL itself by walking activities.db and applying exponential averages, because
// the Ambit3 has no power meter and no decoded HR strap, so its "load" is duration-based
// (see coachservice.cpp's own note: "Load is duration-based (minutes/day) — this device family
// has no power meter or HR strap decoded yet"). intervals.icu already publishes ctl/atl/rampRate
// per day, computed from whatever real TSS each activity carried across every device the athlete
// uses — so reading those is both less code and more accurate than re-deriving a weaker proxy.
//
// The thresholds, the light names and the sentences below are copied exactly from the desktop
// so both apps say the same thing about the same day.

export type ReadinessLight = 'green' | 'tempered' | 'yellow' | 'red';

export interface Readiness {
  fitness: number;      // CTL
  fatigue: number;      // ATL
  freshness: number;    // TSB = CTL - ATL
  rampPerWeek: number;
  light: ReadinessLight;
  sentence: string;
}

export interface ChartPoint {
  date: string;
  fitness: number;
  fatigue: number;
}

// Same wording as the desktop's coachservice.cpp, verbatim.
const SENTENCES: Record<ReadinessLight, string> = {
  green: "You're fresh today — fully rested. Good day for something hard if you've got it planned.",
  tempered: "You're fresh, but your fitness has climbed fast this month. Good day to train — ease into the hard part rather than going in cold.",
  yellow: "You're carrying some fatigue. Nothing alarming — see how the legs feel and adapt if today's plan feels heavy.",
  red: "You're dug in deep right now. Lean toward rest or something gentle today.",
};

// Same cut-offs as the desktop: TSB > -10 green, > -25 yellow, else red; a green day with a
// steep 7-day CTL ramp is downgraded to "tempered" (fresh, but climbing fast).
export function lightFor(freshness: number, rampPerWeek: number): ReadinessLight {
  let light: ReadinessLight = freshness > -10 ? 'green' : freshness > -25 ? 'yellow' : 'red';
  if (light === 'green' && rampPerWeek > 7) light = 'tempered';
  return light;
}

// One recommended session for today, same shape the desktop's pickWorkouts() emits.
export interface WorkoutPick {
  name: string;
  durationSec: number;
  load: number;            // TSS
  intensityFactor: number; // IF
  intensity: string;       // bucket: recovery/endurance/tempo/hard
}

export interface CoachData {
  readiness: Readiness | null;
  chart: ChartPoint[];
  picks: WorkoutPick[];
}

// ── SYSTM catalogue picks — the third of desktop coachservice.cpp that Android was missing.
// Same maths line-for-line as the desktop: bucket by IF, filter by the readiness light, sort
// by TSS, then (André, 2026-08-28: "recommending the same 2 exercises for 3 days in a row")
// exclude anything trained in the last 7 days and rotate the start point by the day so the two
// picks change day-to-day instead of always being the top-TSS pair.

interface CatalogueRow { name: string; tss: number; intensityFactor: number; durationSec: number }

const CATALOGUE: CatalogueRow[] = (systmSample as any[]).map(o => ({
  name: String(o.name ?? ''),
  tss: Number(o.tss ?? 0),
  intensityFactor: Number(o.if ?? 0),
  durationSec: Math.round(Number(o.dur ?? 0)),
}));

// Mirrors systmLibrary.ts bucket() / coachservice.cpp intensityBucket().
function intensityBucket(intensityFactor: number): string {
  if (intensityFactor < 0.60) return 'recovery';
  if (intensityFactor < 0.75) return 'endurance';
  if (intensityFactor < 0.82) return 'tempo';
  return 'hard';
}

// Mirrors coach.ts recommend() / coachservice.cpp intensityForLight().
function intensityForLight(light: ReadinessLight): string {
  if (light === 'red') return 'recovery';
  if (light === 'green') return 'hard';
  return 'endurance';   // tempered / yellow
}

function normalizeName(s: string): string {
  return s.trim().toLowerCase();
}

function pickWorkouts(intensity: string, maxMinutes: number, recentDone: Set<string>, limit = 2): WorkoutPick[] {
  const matched = CATALOGUE
    .filter(r => intensityBucket(r.intensityFactor) === intensity && r.durationSec <= maxMinutes * 60)
    .sort((a, b) => b.tss - a.tss);

  // Drop sessions trained in the last 7 days; if that empties the pool (small catalogue,
  // everything recently done), fall back to the full sorted list rather than showing nothing.
  let pool = matched.filter(r => !recentDone.has(normalizeName(r.name)));
  if (pool.length === 0) pool = matched;

  const n = pool.length;
  if (n === 0) return [];
  // Rotate the window by the day so even when nothing is excluded the two picks still vary
  // from one day to the next. Stable within a day, wraps around the pool. Same julian-style
  // day index the desktop derives from QDate::currentDate().toJulianDay().
  const dayIndex = Math.floor(Date.now() / 86400000);
  const offset = ((dayIndex % n) + n) % n;
  const out: WorkoutPick[] = [];
  for (let k = 0; k < limit && k < n; k++) {
    const r = pool[(offset + k) % n];
    out.push({
      name: r.name, durationSec: r.durationSec, load: r.tss,
      intensityFactor: r.intensityFactor, intensity,
    });
  }
  return out;
}

// Names of activities completed in the last 7 days, for the pick exclusion. Best-effort, same
// honest limitation as the desktop: an activity's name (e.g. "Ride") rarely matches a SYSTM
// session name exactly, so this fires only when they genuinely coincide — never wrongly. The
// rotation above is what guarantees day-to-day variety regardless.
async function recentlyDoneNames(days = 7): Promise<Set<string>> {
  const done = new Set<string>();
  try {
    const creds = await getIntervalsIcuCredentials();
    if (!creds) return done;
    const newest = new Date();
    const oldest = new Date(newest.getTime() - days * 24 * 3600 * 1000);
    const ymd = (d: Date) => d.toISOString().slice(0, 10);
    const url = `https://intervals.icu/api/v1/athlete/${encodeURIComponent(creds.athleteId)}/activities`
      + `?oldest=${ymd(oldest)}&newest=${ymd(newest)}`;
    const resp = await fetch(url, {
      headers: { Authorization: 'Basic ' + btoa(`API_KEY:${creds.apiKey}`), 'User-Agent': 'Sommet/1.0' },
    });
    if (!resp.ok) return done;   // exclusion is a bonus; never let it break the picks
    const acts = await resp.json();
    if (Array.isArray(acts)) {
      for (const a of acts) {
        const norm = normalizeName(String(a?.name ?? ''));
        if (norm) done.add(norm);
      }
    }
  } catch {
    // ignore — fall back to no exclusion
  }
  return done;
}

// `days` bounds the chart window; readiness itself always comes from the most recent day that
// actually has ctl/atl, which is not necessarily today (intervals fills a day in once it has
// processed it).
export async function loadCoachData(days = 90): Promise<CoachData> {
  const rows = await fetchWellnessRaw(days);
  const withLoad = rows.filter(r => typeof r.ctl === 'number' && typeof r.atl === 'number');
  if (withLoad.length === 0) return { readiness: null, chart: [], picks: [] };

  const last = withLoad[withLoad.length - 1];
  const fitness = last.ctl as number;
  const fatigue = last.atl as number;
  const freshness = fitness - fatigue;

  // intervals publishes its own rampRate (CTL change per week). Fall back to measuring the
  // 7-day CTL delta ourselves when it is absent, which is what the desktop does.
  let rampPerWeek = typeof last.rampRate === 'number' ? last.rampRate : 0;
  if (typeof last.rampRate !== 'number' && withLoad.length > 7) {
    const prev = withLoad[withLoad.length - 8];
    rampPerWeek = fitness - (prev.ctl as number);
  }

  const light = lightFor(freshness, rampPerWeek);
  const recentDone = await recentlyDoneNames(7);
  const picks = pickWorkouts(intensityForLight(light), 90, recentDone);
  return {
    readiness: { fitness, fatigue, freshness, rampPerWeek, light, sentence: SENTENCES[light] },
    chart: withLoad.map(r => ({ date: r.date, fitness: r.ctl as number, fatigue: r.atl as number })),
    picks,
  };
}

// WellnessService.fetchWellness() deliberately narrows to the health metrics its two screens
// need and drops ctl/atl/rampRate. Rather than widen that shared shape for one caller, this
// reads the same endpoint and keeps the training-load fields instead.
interface LoadDay { date: string; ctl?: number; atl?: number; rampRate?: number }

async function fetchWellnessRaw(days: number): Promise<LoadDay[]> {
  const creds = await getIntervalsIcuCredentials();
  if (!creds) return [];

  const newest = new Date();
  const oldest = new Date(newest.getTime() - days * 24 * 3600 * 1000);
  const ymd = (d: Date) => d.toISOString().slice(0, 10);
  const url = `https://intervals.icu/api/v1/athlete/${encodeURIComponent(creds.athleteId)}/wellness`
    + `?oldest=${ymd(oldest)}&newest=${ymd(newest)}`;
  const resp = await fetch(url, {
    headers: { Authorization: 'Basic ' + btoa(`API_KEY:${creds.apiKey}`), 'User-Agent': 'Sommet/1.0' },
  });
  if (!resp.ok) throw new Error(`intervals.icu wellness: HTTP ${resp.status}`);
  const rows = await resp.json();
  if (!Array.isArray(rows)) return [];

  return rows
    .map((r: any) => ({
      date: String(r?.id ?? ''),
      ctl: typeof r?.ctl === 'number' ? r.ctl : undefined,
      atl: typeof r?.atl === 'number' ? r.atl : undefined,
      rampRate: typeof r?.rampRate === 'number' ? r.rampRate : undefined,
    }))
    .filter(r => r.date)
    .sort((a, b) => a.date.localeCompare(b.date));
}

File: desktop/src/services/coachservice.cpp
#include "coachservice.h"

#include <QDateTime>
#include <QDir>
#include <QFile>
#include <QJsonArray>
#include <QJsonDocument>
#include <QJsonObject>
#include <QMap>
#include <QNetworkReply>
#include <QNetworkRequest>
#include <QSqlError>
#include <QSqlQuery>
#include <QStandardPaths>
#include <QUrl>
#include <QVector>
#include <algorithm>
#include <cmath>

namespace {
// Same maths as coach/src/coach.ts and the Fresh Today / Ride Coach mockups - k=42 for
// Fitness, k=7 for Fatigue, the standard exponentially-weighted Performance-Management
// formulation. Kept as free functions (not methods) so there's exactly one place this can
// drift from the TypeScript original.
double alphaCtl() { return 1.0 - std::exp(-1.0 / 42.0); }
double alphaAtl() { return 1.0 - std::exp(-1.0 / 7.0); }
}

CoachService::CoachService(QObject *parent) : QObject(parent)
{
    m_settings.beginGroup(QStringLiteral("coach"));
    m_chatBackend = m_settings.value(QStringLiteral("chatBackend"), QStringLiteral("canned")).toString();
    m_catalogueSource = m_settings.value(QStringLiteral("catalogueSource"), QStringLiteral("sample")).toString();
    m_systmMcpUrl = m_settings.value(QStringLiteral("systmMcpUrl")).toString();
    m_settings.endGroup();

    openActivitiesDb();

    // Bundled catalogue, read once via Qt's compiled-in resource system (see CMakeLists.txt
    // RESOURCES) - not a filesystem path, so it works the same in a packaged build as it
    // does here.
    QFile f(QStringLiteral(":/qt/qml/AmbitApp/assets/coach/systm-sample.json"));
    if (f.open(QIODevice::ReadOnly)) {
        const auto doc = QJsonDocument::fromJson(f.readAll());
        for (const auto &v : doc.array()) {
            const auto o = v.toObject();
            m_sampleCatalogue.append({o.value(QStringLiteral("name")).toString(),
                                       o.value(QStringLiteral("tss")).toDouble(),
                                       o.value(QStringLiteral("if")).toDouble(),
                                       int(o.value(QStringLiteral("dur")).toDouble())});
        }
    }

    computeReadiness();
    // Opening line, same as the mockup's apply('rested') seeding the transcript on load.
    appendBubble(QStringLiteral("coach"),
                 m_readiness.value(QStringLiteral("sentence")).toString(), pickWorkouts(
                     intensityForLight(m_readiness.value(QStringLiteral("light")).toString()), 90));

    if (m_catalogueSource == QStringLiteral("live") && !m_systmMcpUrl.isEmpty()) {
        refreshCatalogueLive([]() {});
    }
}

void CoachService::openActivitiesDb()
{
    // Named connection distinct from ActivityService's own "activities" connection - same
    // file, read-only use here, but Qt's addDatabase() silently steals a connection name
    // reused across classes (ActivityService.h's own header comment already documents this
    // exact trap).
    m_db = QSqlDatabase::addDatabase(QStringLiteral("QSQLITE"), QStringLiteral("coach_activities"));
    const QString dir = QStandardPaths::writableLocation(QStandardPaths::AppDataLocation);
    m_db.setDatabaseName(dir + QStringLiteral("/activities.db"));
    // Deliberately NOT creating the table here - ActivityService owns that. If it hasn't
    // run yet (fresh install, no watch ever synced), the SELECT below just finds no table
    // and computeReadiness() falls back to its own "no history yet" default.
    m_db.open();
}

void CoachService::computeReadiness()
{
    const QDate today = QDate::currentDate();
    const QDate doneSince = today.addDays(-7);   // "recently done" window for pickWorkouts()
    m_recentDone.clear();

    QMap<QDate, double> loadByDay;   // day -> minutes trained that day (the load proxy)
    if (m_db.isOpen()) {
        QSqlQuery q(QStringLiteral(
            "SELECT start_time, duration_s, name FROM activities "
            "WHERE start_time IS NOT NULL AND start_time != ''"), m_db);
        while (q.next()) {
            const QDateTime dt = QDateTime::fromString(q.value(0).toString(), Qt::ISODate);
            if (!dt.isValid()) continue;
            const double minutes = q.value(1).toDouble() / 60.0;
            loadByDay[dt.date()] += minutes;
            // Anything trained in the last 7 days is a candidate for exclusion from picks.
            // Name-matching is best-effort: a ride recorded on the watch carries its sport-mode
            // name, not the SYSTM session name, so this only fires when the two genuinely match
            // (e.g. a workout named the same way) — but it never wrongly excludes.
            if (dt.date() >= doneSince) {
                const QString norm = normalizeName(q.value(2).toString());
                if (!norm.isEmpty()) m_recentDone.insert(norm);
            }
        }
    }

    QDate start = today.addDays(-119);   // cap history depth - bounded, still plenty for a 42d ramp
    if (!loadByDay.isEmpty()) {
        const QDate earliest = loadByDay.firstKey();
        if (earliest > start) start = earliest;
    } else {
        start = today;   // no history at all - a single, flat "day"
    }

    QVector<QDate> days;
    QVector<double> ctlArr, atlArr, freshArr;
    double ctl = 0.0, atl = 0.0;
    const double ac = alphaCtl(), aa = alphaAtl();
    for (QDate d = start; d <= today; d = d.addDays(1)) {
        days.append(d);
        freshArr.append(ctl - atl);            // freshness AS OF THE START of this day
        const double load = loadByDay.value(d, 0.0);
        ctl = ctl * (1 - ac) + load * ac;
        atl = atl * (1 - aa) + load * aa;
        ctlArr.append(ctl);
        atlArr.append(atl);
    }

    const int n = days.size();
    const double finalCtl = ctlArr.last(), finalAtl = atlArr.last(), finalTsb = freshArr.last();

    double rampPerWeek = 0.0;
    for (int d = n - 1; d >= qMax(7, n - 28); --d)
        rampPerWeek = qMax(rampPerWeek, ctlArr[d] - ctlArr[d - 7]);

    QString light = finalTsb > -10 ? QStringLiteral("green")
                  : finalTsb > -25 ? QStringLiteral("yellow") : QStringLiteral("red");
    if (light == QStringLiteral("green") && rampPerWeek > 7) light = QStringLiteral("tempered");

    QString sentence;
    if (light == QStringLiteral("green"))
        sentence = QStringLiteral("You're fresh today — fully rested. Good day for something hard if you've got it planned.");
    else if (light == QStringLiteral("tempered"))
        sentence = QStringLiteral("You're fresh, but your fitness has climbed fast this month. Good day to train — ease into the hard part rather than going in cold.");
    else if (light == QStringLiteral("yellow"))
        sentence = QStringLiteral("You're carrying some fatigue. Nothing alarming — see how the legs feel and adapt if today's plan feels heavy.");
    else
        sentence = QStringLiteral("You're dug in deep right now. Lean toward rest or something gentle today.");

    m_readiness = QVariantMap{
        {QStringLiteral("fitness"), finalCtl}, {QStringLiteral("fatigue"), finalAtl},
        {QStringLiteral("freshness"), finalTsb}, {QStringLiteral("rampPerWeek"), rampPerWeek},
        {QStringLiteral("light"), light}, {QStringLiteral("sentence"), sentence},
        {QStringLiteral("basis"), QVariantList{QStringLiteral("load")}},
    };

    m_chartSeries.clear();
    const int keep = qMin(42, n);
    for (int i = n - keep; i < n; ++i) {
        m_chartSeries.append(QVariantMap{
            {QStringLiteral("date"), days[i].toString(Qt::ISODate)},
            {QStringLiteral("fitness"), ctlArr[i]}, {QStringLiteral("fatigue"), atlArr[i]},
        });
    }

    m_todaysPicks = pickWorkouts(intensityForLight(light), 90);
    emit readinessChanged();
}

QString CoachService::intensityBucket(double intensityFactor)
{
    // Mirrors coach/src/adapters/systmLibrary.ts's bucket() exactly - SYSTM has no single
    // "intensity" field in this offline sample, so it's derived from IF the same way there.
    if (intensityFactor < 0.60) return QStringLiteral("recovery");
    if (intensityFactor < 0.75) return QStringLiteral("endurance");
    if (intensityFactor < 0.82) return QStringLiteral("tempo");
    return QStringLiteral("hard");
}

QString CoachService::intensityForLight(const QString &light)
{
    // Mirrors coach/src/coach.ts's recommend(): red -> recovery, green -> hard, everything
    // else (tempered/yellow) -> endurance.
    if (light == QStringLiteral("red")) return QStringLiteral("recovery");
    if (light == QStringLiteral("green")) return QStringLiteral("hard");
    return QStringLiteral("endurance");
}

QString CoachService::normalizeName(const QString &s)
{
    return s.trimmed().toLower();
}

QVariantList CoachService::pickWorkouts(const QString &intensity, int maxMinutes, int limit) const
{
    const QList<CatalogueRow> &rows =
        (m_catalogueSource == QStringLiteral("live") && !m_liveCatalogue.isEmpty())
            ? m_liveCatalogue : m_sampleCatalogue;

    QList<CatalogueRow> matched;
    for (const auto &r : rows) {
        if (intensityBucket(r.intensityFactor) != intensity) continue;
        if (r.durationSec > maxMinutes * 60) continue;
        matched.append(r);
    }
    std::sort(matched.begin(), matched.end(),
              [](const CatalogueRow &a, const CatalogueRow &b) { return a.tss > b.tss; });

    // Drop anything trained in the last 7 days so the coach stops re-serving a session you've
    // just done (André, 2026-08-28: "recommending the same 2 exercises for 3 days in a row").
    // If that would empty the pool (small catalogue, everything recently done), fall back to
    // the full sorted list rather than showing nothing.
    QList<CatalogueRow> pool;
    for (const auto &r : matched)
        if (!m_recentDone.contains(normalizeName(r.name))) pool.append(r);
    if (pool.isEmpty()) pool = matched;

    // Rotate the starting point by the day, so even when nothing gets excluded the two picks
    // still change from one day to the next instead of always being the top-TSS pair. Stable
    // within a single day (same julian-day offset), varied across days, wraps around the pool.
    QVariantList out;
    const int n = pool.size();
    if (n == 0) return out;
    const int offset = int(QDate::currentDate().toJulianDay() % n);
    for (int k = 0; k < limit && k < n; ++k) {
        const auto &r = pool[(offset + k) % n];
        out.append(QVariantMap{
            {QStringLiteral("name"), r.name}, {QStringLiteral("durationSec"), r.durationSec},
            {QStringLiteral("load"), r.tss}, {QStringLiteral("intensityFactor"), r.intensityFactor},
            {QStringLiteral("intensity"), intensity},
        });
    }
    return out;
}

void CoachService::refreshCatalogueLive(std::function<void()> onDone)
{
    // Honest limitation (see this class's own header comment): joaodrp/wahoo-systm-mcp
    // speaks MCP over stdio, not plain HTTP - this GETs systmMcpUrl expecting a small bridge
    // in front of it that returns the same JSON array shape as the bundled sample
    // ([{name, tss, if, dur, steps}]). Real code; needs that bridge actually running.
    QNetworkRequest req{QUrl(m_systmMcpUrl)};
    auto *reply = m_net.get(req);
    connect(reply, &QNetworkReply::finished, this, [this, reply, onDone]() {
        reply->deleteLater();
        if (reply->error() != QNetworkReply::NoError) {
            setLastError(QStringLiteral("live catalogue fetch failed: %1").arg(reply->errorString()));
            onDone();
            return;
        }
        const auto doc = QJsonDocument::fromJson(reply->readAll());
        QList<CatalogueRow> rows;
        for (const auto &v : doc.array()) {
            const auto o = v.toObject();
            rows.append({o.value(QStringLiteral("name")).toString(),
                         o.value(QStringLiteral("tss")).toDouble(),
                         o.value(QStringLiteral("if")).toDouble(),
                         int(o.value(QStringLiteral("dur")).toDouble())});
        }
        m_liveCatalogue = rows;
        computeReadiness();   // todaysPicks may change with the new catalogue
        onDone();
    });
}

void CoachService::refreshReadiness() { computeReadiness(); }

void CoachService::appendBubble(const QString &role, const QString &text, const QVariantList &cards)
{
    m_messages.append(QVariantMap{{QStringLiteral("role"), role}, {QStringLiteral("text"), text},
                                   {QStringLiteral("cards"), cards}});
    emit messagesChanged();
}

void CoachService::resetConversation()
{
    m_messages.clear();
    appendBubble(QStringLiteral("coach"),
                 m_readiness.value(QStringLiteral("sentence")).toString(),
                 pickWorkouts(intensityForLight(m_readiness.value(QStringLiteral("light")).toString()), 90));
}

void CoachService::sendMessage(const QString &text)
{
    const QString trimmed = text.trimmed();
    if (trimmed.isEmpty()) return;
    appendBubble(QStringLiteral("me"), trimmed);
    setSending(true);
    if (m_chatBackend == QStringLiteral("claude") && anthropicKeySet()) {
        replyClaude(trimmed);
    } else {
        replyCanned(trimmed);
    }
}

void CoachService::replyCanned(const QString &userText)
{
    const QString t = userText.toLower();
    const QString light = m_readiness.value(QStringLiteral("light")).toString();
    const QString intensity = intensityForLight(light);

    if (t.contains(QStringLiteral("short"))) {
        appendBubble(QStringLiteral("coach"), QStringLiteral("Sure — trimmed it down:"),
                     pickWorkouts(intensity, 30));
    } else if (t.contains(QStringLiteral("outdoor"))) {
        appendBubble(QStringLiteral("coach"),
            QStringLiteral("I don't have live weather wired into this reply yet — check the "
                           "forecast on Home, then pick a route from Routes and go."));
    } else if (t.contains(QStringLiteral("watch")) || t.contains(QStringLiteral("send"))) {
        const auto picks = m_todaysPicks;
        const QString name = picks.isEmpty() ? QStringLiteral("today's session")
                                              : picks.first().toMap().value(QStringLiteral("name")).toString();
        appendBubble(QStringLiteral("coach"),
            QStringLiteral("Sending \"%1\" to your watch isn't wired up from this page yet — "
                           "Ambit3Sink exists in the coach scaffold (planned-move + App-Zone "
                           "routes), it just isn't called from here.").arg(name));
    } else {
        appendBubble(QStringLiteral("coach"),
            QStringLiteral("I'm running in canned mode right now (Settings → Coach → Chat "
                           "backend, switch to Claude for a real conversation). I only "
                           "recognise a few keywords here: try \"something shorter\", "
                           "\"outdoor instead\", or \"send it to my watch\"."));
    }
    setSending(false);
}

void CoachService::replyClaude(const QString &userText)
{
    const QVariantMap r = m_readiness;
    QStringList picksLines;
    for (const auto &pv : m_todaysPicks) {
        const auto p = pv.toMap();
        picksLines << QStringLiteral("- %1 (%2 min, TSS %3)")
            .arg(p.value(QStringLiteral("name")).toString())
            .arg(p.value(QStringLiteral("durationSec")).toInt() / 60)
            .arg(p.value(QStringLiteral("load")).toDouble(), 0, 'f', 0);
    }
    // No onboarding UI built yet for the rider profile coach/src/model.ts defines
    // (mainSports/secondarySports/weatherMatters/recoveryMenu) - honestly a placeholder
    // until that screen exists, not fabricated as if it were real user input.
    const QString system = QStringLiteral(
        "You are Sommet's training coach for a Suunto Ambit3 watch user. Be concise and warm, "
        "like a knowledgeable training partner, not a corporate assistant. Never use jargon "
        "(no \"TSS\"/\"CTL\"/\"TSB\" in your reply - translate to plain language).\n\n"
        "Today's readiness: %1 (%2). Fitness %3, Fatigue %4, Freshness %5.\n"
        "Today's catalogue picks for this readiness:\n%6\n\n"
        "Rider profile: not yet collected by this app (no onboarding screen built) - don't "
        "assume specifics about their sport or goals beyond what they tell you in the chat."
    ).arg(r.value(QStringLiteral("light")).toString(), r.value(QStringLiteral("sentence")).toString())
     .arg(r.value(QStringLiteral("fitness")).toDouble(), 0, 'f', 0)
     .arg(r.value(QStringLiteral("fatigue")).toDouble(), 0, 'f', 0)
     .arg(r.value(QStringLiteral("freshness")).toDouble(), 0, 'f', 0)
     .arg(picksLines.isEmpty() ? QStringLiteral("(none match today's duration/intensity)")
                                : picksLines.join(QStringLiteral("\n")));

    QNetworkRequest req{QUrl(QStringLiteral("https://api.anthropic.com/v1/messages"))};
    req.setHeader(QNetworkRequest::ContentTypeHeader, QStringLiteral("application/json"));
    req.setRawHeader("x-api-key", m_settings.value(QStringLiteral("coach/anthropicApiKey")).toString().toUtf8());
    req.setRawHeader("anthropic-version", "2023-06-01");

    QJsonObject body;
    body[QStringLiteral("model")] = QStringLiteral("claude-sonnet-5");
    body[QStringLiteral("max_tokens")] = 500;
    body[QStringLiteral("system")] = system;
    QJsonArray messages;
    QJsonObject userMsg;
    userMsg[QStringLiteral("role")] = QStringLiteral("user");
    userMsg[QStringLiteral("content")] = userText;
    messages.append(userMsg);
    body[QStringLiteral("messages")] = messages;

    auto *reply = m_net.post(req, QJsonDocument(body).toJson(QJsonDocument::Compact));
    connect(reply, &QNetworkReply::finished, this, [this, reply]() {
        reply->deleteLater();
        if (reply->error() != QNetworkReply::NoError) {
            setLastError(QStringLiteral("Claude API error: %1").arg(reply->errorString()));
            appendBubble(QStringLiteral("coach"),
                QStringLiteral("(couldn't reach Claude: %1 — falling back, try again in a moment)")
                    .arg(reply->errorString()));
            setSending(false);
            return;
        }
        const auto doc = QJsonDocument::fromJson(reply->readAll());
        const auto content = doc.object().value(QStringLiteral("content")).toArray();
        QString text;
        if (!content.isEmpty()) text = content.first().toObject().value(QStringLiteral("text")).toString();
        if (text.isEmpty()) text = QStringLiteral("(empty reply from Claude)");
        appendBubble(QStringLiteral("coach"), text);
        setSending(false);
    });
}

void CoachService::setChatBackend(const QString &v)
{
    if (m_chatBackend == v) return;
    m_chatBackend = v;
    m_settings.setValue(QStringLiteral("coach/chatBackend"), v);
    emit chatBackendChanged();
}

void CoachService::setCatalogueSource(const QString &v)
{
    if (m_catalogueSource == v) return;
    m_catalogueSource = v;
    m_settings.setValue(QStringLiteral("coach/catalogueSource"), v);
    emit catalogueSourceChanged();
    if (v == QStringLiteral("live") && !m_systmMcpUrl.isEmpty()) {
        refreshCatalogueLive([]() {});
    } else {
        computeReadiness();
    }
}

bool CoachService::anthropicKeySet() const
{
    return !m_settings.value(QStringLiteral("coach/anthropicApiKey")).toString().isEmpty();
}

void CoachService::setAnthropicApiKey(const QString &key)
{
    m_settings.setValue(QStringLiteral("coach/anthropicApiKey"), key.trimmed());
    emit anthropicKeySetChanged();
}

void CoachService::clearAnthropicApiKey()
{
    m_settings.remove(QStringLiteral("coach/anthropicApiKey"));
    emit anthropicKeySetChanged();
}

void CoachService::setSystmMcpUrl(const QString &v)
{
    if (m_systmMcpUrl == v) return;
    m_systmMcpUrl = v;
    m_settings.setValue(QStringLiteral("coach/systmMcpUrl"), v);
    emit systmMcpUrlChanged();
    if (m_catalogueSource == QStringLiteral("live") && !v.isEmpty()) {
        refreshCatalogueLive([]() {});
    }
}

void CoachService::setSending(bool v)
{
    if (m_sending == v) return;
    m_sending = v;
    emit sendingChanged();
}

void CoachService::setLastError(const QString &e)
{
    m_lastError = e;
    emit lastErrorChanged();
}
